django/db/models/indexes.py

```diff
@@ -0,0 +1,120 @@
+            self._name = self.get_name()
+            self.normalise_name()
+        return self._name
+
+    def normalise_name(self):
+        # Name shouldn't start with an underscore (Oracle hates this), so prepend D if we need to
+        if self._name[0] == "_":
+            self._name = 'D%s' % self._name[1:]
+        # It can't start with a number on Oracle, so prepend D if we need to
+        elif self._name[0].isdigit():
+            self._name = 'D%s' % self._name[1:]
+
+    def create_sql(self, schema_editor):
+        columns = [field for field in self.fields]
+        fields = [self.model._meta.get_field(field) for field in self.fields]
+        if len(fields) == 1 and fields[0].db_tablespace:
+            tablespace_sql = schema_editor.connection.ops.tablespace_sql(fields[0].db_tablespace)
+        elif self.model._meta.db_tablespace:
+            tablespace_sql = schema_editor.connection.ops.tablespace_sql(self.model._meta.db_tablespace)
+        else:
+            tablespace_sql = ""
+        if tablespace_sql:
+            tablespace_sql = " " + tablespace_sql
+
+        quote_name = schema_editor.quote_name
+        return schema_editor.sql_create_index % {
+            'table': quote_name(self.model._meta.db_table),
+            'name': quote_name(self.name),
+            'columns': ', '.join(quote_name(column) for column in columns),
+            'extra': tablespace_sql,
+        }
+
+    def remove_sql(self, schema_editor):
+        quote_name = schema_editor.quote_name
+        return schema_editor.sql_delete_index % {
+            'table': quote_name(self.model._meta.db_table),
+            'name': quote_name(self.name),
+        }
+
+    def deconstruct(self):
+        """
+        Return a 3-tuple of class import path, positional arguments, and
+        keyword arguments.
+        """
+        path = "%s.%s" % (self.__class__.__module__, self.__class__.__name__)
+        path = path.replace("django.db.models.indexes", "django.db.models")
+        return (path, (), {'fields': self.fields})
+
+    @staticmethod
+    def _hash_generator(*args):
+        """
+        Generate a 32-bit digest of a set of arguments that can be used to
+        shorten identifying names.
+        """
+        h = hashlib.md5()
+        for arg in args:
+            h.update(force_bytes(arg))
+        return h.hexdigest()[:6]
+
+    def get_name(self, suffix='_idx'):
+        """
+        Generate a unique name for the index.
+
+        The name is divided into 3 parts - table name (12 chars), field name
+        (8 chars) and unique hash + suffix (10 chars). Each part is made to
+        fit its size by truncating the excess length.
+        """
+        table_name = self.model._meta.db_table
+        column_names = self.fields
+        hash_data = [table_name] + column_names + [self.index_type]
+        index_unique_hash = self._hash_generator(*hash_data)
+        table_part = table_name.replace('"', '').replace('.', '_')
```

Comment:
Sorry but even I am not sure what these replacements are for. They were introduced [here](https://github.com/django/django/commit/b139315f1c5e3eb05c76237c2824bdf03bd689b6#diff-bdef24317b25b8cdb404f4e09be13966R401) and I have been unable to find any tests that are/were related to them. Also, removing them breaks migrations when I had a `db_table='name"with'` in my `Meta` with the error
> Django.db.utils.OperationalError: near "with": syntax error